File: backend/app/integrations/github/checks_publisher.py
from app.integrations.github.client import GitHubClient


class ChecksPublisher:
    """GitHub checks/status publisher with shared HTTP client."""

    def __init__(self, github_token: str):
        self.github_token = github_token
        self._client = GitHubClient(github_token)

    def create_check_run(self, owner: str, repo: str, *, name: str, head_sha: str, details_url: str | None = None) -> int:
        url = f'/repos/{owner}/{repo}/check-runs'
        payload = {'name': name, 'head_sha': head_sha, 'status': 'in_progress', 'output': {'title': name, 'summary': 'SecurePR AI is running.'}}
        if details_url:
            payload['details_url'] = details_url
        r = self._client.http.post(url, json=payload)
        return int(r.json().get('id'))

    def update_check_run(self, owner: str, repo: str, check_run_id: int, *, conclusion: str, summary: str, details_url: str | None = None):
        url = f'/repos/{owner}/{repo}/check-runs/{check_run_id}'
        payload = {'status': 'completed', 'conclusion': conclusion, 'output': {'title': 'SecurePR AI', 'summary': summary}}
        if details_url:
            payload['details_url'] = details_url
        r = self._client.http.patch(url, json=payload)
        return r.json()

    def create_commit_status(self, owner: str, repo: str, sha: str, *, state: str, context: str, description: str = '', target_url: str | None = None):
        url = f'/repos/{owner}/{repo}/statuses/{sha}'
        payload = {'state': state, 'context': context}
        if description:
            payload['description'] = description
        if target_url:
            payload['target_url'] = target_url
        r = self._client.http.post(url, json=payload)
        return r.json()

File: backend/src/integrations/github/checks-publisher.ts
import { GitHubClient } from './client.js';

/**
 * GitHub checks/status publisher with shared HTTP client.
 */
export class ChecksPublisher {
  private client: GitHubClient;

  constructor(githubToken: string) {
    this.client = GitHubClient.getInstance(githubToken);
  }

  async createCheckRun(
    owner: string, repo: string,
    options: { name: string; headSha: string; detailsUrl?: string }
  ): Promise<number> {
    const url = `/repos/${owner}/${repo}/check-runs`;
    const payload: Record<string, unknown> = {
      name: options.name,
      head_sha: options.headSha,
      status: 'in_progress',
      output: { title: options.name, summary: 'SecurePR AI is running.' },
    };
    if (options.detailsUrl) {
      payload.details_url = options.detailsUrl;
    }
    const r = await this.client.http.post(url, { json: payload });
    return parseInt(r.data.id, 10);
  }

  async updateCheckRun(
    owner: string, repo: string, checkRunId: number,
    options: { conclusion: string; summary: string; detailsUrl?: string }
  ): Promise<Record<string, unknown>> {
    const url = `/repos/${owner}/${repo}/check-runs/${checkRunId}`;
    const payload: Record<string, unknown> = {
      status: 'completed',
      conclusion: options.conclusion,
      output: { title: 'SecurePR AI', summary: options.summary },
    };
    if (options.detailsUrl) {
      payload.details_url = options.detailsUrl;
    }
    const r = await this.client.http.patch(url, { json: payload });
    return r.data;
  }

  async createCommitStatus(
    owner: string, repo: string, sha: string,
    options: { state: string; context: string; description?: string; targetUrl?: string }
  ): Promise<Record<string, unknown>> {
    const url = `/repos/${owner}/${repo}/statuses/${sha}`;
    const payload: Record<string, unknown> = {
      state: options.state,
      context: options.context,
    };
    if (options.description) {
      payload.description = options.description;
    }
    if (options.targetUrl) {
      payload.target_url = options.targetUrl;
    }
    const r = await this.client.http.post(url, { json: payload });
    return r.data;
  }
}
